module/os/linux/zfs/vdev_disk.c

```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
 }
 
 static inline unsigned int
-vdev_bio_max_segs(zio_t *zio, int bio_size, uint64_t abd_offset)
+vdev_bio_max_segs(struct block_device *bdev)
 {
-	unsigned long nr_segs = abd_nr_pages_off(zio->io_abd,
-	    bio_size, abd_offset);
+	/*
+	 * Smallest of the device max segs and the tuneable max segs. Minimum
+	 * 4, so there's room to finish split pages if they come up.
+	 */
+	const unsigned long dev_max_segs =
+	    queue_max_segments(bdev_get_queue(bdev));
+	const unsigned long tune_max_segs = (vdev_disk_max_segs > 0) ?
+	    MAX(4, vdev_disk_max_segs) : dev_max_segs;
+	const unsigned long max_segs = MIN(tune_max_segs, dev_max_segs);
 
 #ifdef HAVE_BIO_MAX_SEGS
-	return (bio_max_segs(nr_segs));
+	return (bio_max_segs(max_segs));
 #else
-	return (MIN(nr_segs, BIO_MAX_PAGES));
+	return (MIN(max_segs, BIO_MAX_PAGES));
 #endif
 }
 
+
+/*
+ * Virtual block IO object (VBIO)
+ *
+ * Linux block IO (BIO) objects have a limit on how many data segments (pages)
+ * they can hold. Depending on how they're allocated and structured, a large
+ * ZIO can require more than one BIO to be submitted to the kernel, which then
+ * all have to complete before we can return the completed ZIO back to ZFS.
+ *
+ * A VBIO is a wrapper around multiple BIOs, carrying everything needed to
+ * translate a ZIO down into the kernel block layer and back again.
+ *
+ * Note that these are only used for data ZIOs (read/write). Meta-operations
+ * (flush/trim) don't need multiple BIOs and so can just make the call
+ * directly.
+ */
+typedef struct {
+	zio_t		*vbio_zio;	/* parent zio */
+
+	struct block_device *vbio_bdev;	/* blockdev to submit bios to */
+
+	abd_t		*vbio_abd;	/* abd carrying borrowed linear buf */
+
+	atomic_t	vbio_ref;	/* bio refcount */
+	int		vbio_error;	/* error from failed bio */
+
+	uint_t		vbio_max_segs;	/* max segs per bio */
+	uint_t		vbio_max_bios;	/* max bios (size of vbio_bio) */
+
+	uint64_t	vbio_offset;	/* start offset of next bio */
+
+	uint_t		vbio_nbios;	/* allocated bios */
+	uint_t		vbio_cur;	/* current bio */
+	struct bio	*vbio_bio[];	/* attached bios */
+} vbio_t;
+
+static vbio_t *
+vbio_alloc(zio_t *zio, struct block_device *bdev, unsigned int pages)
+{
+	unsigned int max_segs = vdev_bio_max_segs(bdev);
+	int max_bios = (pages / max_segs)+1;
```

Comment:
The math was slightly off here, its true, but you got me thinking more about logical block sizes and yes, I hadn't managed it properly.

In another part of the code, just pushed, I now get the logical block size from the device, and use that when checking the alignment - instead of making sure they're all page aligned, I instead make sure they're all LBS aligned. Really only matters for the head and tail parts anyway, I think.

Linux up to 6.7 restricts logical block sizes to one of 512, 1024, 2048 or 4096 bytes.